src/HaskellCI/GitHub.hs

```diff
@@ -98,6 +100,14 @@ makeGitHub _argv config@Config {..} gitconfig prj jobs@JobVersions {..} = do
             [ ("CC", "${{ matrix.compiler }}")
             ]
 
+    -- Validity checks
+    when (cfgSubmodules && cfgUbuntu < Focal) $
+        throwErr $ ShError $ "Using submodules on the GitHub Actions backend requires "
+                          ++ "Ubuntu 20.04 (Focal Fossa) or later."
+
+    when (anyGHCJS && cfgUbuntu > Bionic) $
```

Comment:
This check is general, i.e. would apply to travis and bash backends too.

OTOH, the checkout nonsense is GHA specific.